codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -214,7 +232,71 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
             dstPort = header.readUnsignedShort();
         }
 
-        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort);
+        final List<HAProxyTLV> tlvs = readTlvs(header);
+
+        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort, tlvs);
+    }
+
+    private static List<HAProxyTLV> readTlvs(final ByteBuf header) {
+        HAProxyTLV haProxyTLV = readNextTLV(header);
+        if (haProxyTLV == null) {
+            return Collections.emptyList();
+        }
+        // In most cases there are less than 4 TLVs available
+        List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>(4);
+
+        do {
+            haProxyTLVs.add(haProxyTLV);
+            if (haProxyTLV instanceof HAProxySSLTLV) {
+                haProxyTLVs.addAll(((HAProxySSLTLV) haProxyTLV).encapsulatedTLVs());
+            }
+        } while ((haProxyTLV = readNextTLV(header)) != null);
+        return haProxyTLVs;
+    }
+
+    private static HAProxyTLV readNextTLV(final ByteBuf header) {
+
+        // We need at least 4 bytes for a TLV
+        if (header.readableBytes() < 4) {
+            return null;
+        }
+
+        final byte typeAsByte = header.readByte();
+        final HAProxyTLV.Type type = HAProxyTLV.Type.typeForByteValue(typeAsByte);
+
+        final int length = header.readUnsignedShort();
+        switch (type) {
+        case PP2_TYPE_SSL:
+            final ByteBuf rawContent = header.retainedSlice(header.readerIndex(), length);
+            final ByteBuf byteBuf = header.readSlice(length);
+            final byte client = byteBuf.readByte();
+            final int verify = byteBuf.readInt();
+
+            if (byteBuf.readableBytes() >= 4) {
+
+                final List<HAProxyTLV> encapsulatedTlvs = new ArrayList<HAProxyTLV>(4);
+                while (byteBuf.readableBytes() >= 4) {
```

Comment:
you have already checked above `if (byteBuf.readableBytes() >= 4)` ... consider making this a do/while.